### PR TITLE
New Flatpak CI and Manifest

### PR DESCRIPTION
**Pull request by istori1**
## Description
Changes the CI and Flatpak manifest for easier merging of future changes on the flathub repo.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [x] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

### CODE REVIEW

File: packaging/linux/flatpak/dev.lizardbyte.sunshine.yml [hunk 1]
```diff
@@ -98,75 +99,56 @@ modules:
...
+          url-template: https://www.x.org/archive/individual/app/xrandr-$version.tar.xz
+
+  - name: intel-mediasdk
+    only-arches:
+      - x86_64
+    buildsystem: cmake-ninja
+    config-opts:
+      - -DBUILD_SAMPLES=OFF
+      - -DBUILD_TUTORIALS=OFF
+      - -DENABLE_X11=OFF
```
Comment (ReenigneArcher) [inline packaging/linux/flatpak/dev.lizardbyte.sunshine.yml:142]:
```suggestion
    config-opts:
      - -DENABLE_OPENCL=ON
      - -DENABLE_X11_DRI3=ON
      - -DENABLE_WAYLAND=ON
      - -DENABLE_ITT=OFF
      - -DENABLE_TEXTLOG=OFF
      - -DENABLE_STAT=OFF
      - -DBUILD_ALL=OFF
      - -DBUILD_RUNTIME=ON
      - -DBUILD_SAMPLES=OFF
      - -DBUILD_TESTS=OFF
      - -DBUILD_TOOLS=OFF
      - -DUSE_SYSTEM_GTEST=OFF
      - -DMFX_ENABLE_KERNELS=ON
```

I explicitly called out everything because there is no guarantee the defaults will remain the same in future versions.

File: packaging/linux/flatpak/dev.lizardbyte.sunshine.yml [hunk 2]
```diff
@@ -98,75 +99,56 @@ modules:
...
+          type: anitya
+          project-id: 2507
+          stable-only: true
+          url-template: https://github.com/numactl/numactl/archive/refs/tags/v$version.tar.gz
 
-  - name: numactl
-    buildsystem: autotools
-    make-args:
-      - install
+  - name: xrandr
```
Comment (ReenigneArcher) [inline packaging/linux/flatpak/dev.lizardbyte.sunshine.yml:124]: This is probably kind of dumb to include just because of the example app that changes resolution. I'd prefer to just remove the "low res desktop" option from the `apps.json`.

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -218,20 +213,46 @@ jobs:
             -DSUNSHINE_CONFIGURE_ONLY=ON \
             ..
 
+      - name: Generate cache for npm
+        run: |
+          npm install --lockfile-version 2 --package-lock-only
+          git clone --depth 1 https://github.com/flatpak/flatpak-builder-tools.git
```
Comment (ReenigneArcher) [inline .github/workflows/CI.yml:219]: Use github checkout action

File: .github/workflows/CI.yml [hunk 4]
```diff
@@ -218,20 +213,46 @@ jobs:
...
+          pip install ./flatpak-builder-tools/node
+          flatpak-node-generator -o npm-generated-sources.json npm package-lock.json
+          cp npm-generated-sources.json build/npm-generated-sources.json
+          mv npm-generated-sources.json artifacts/npm-generated-sources.json
+          mv package-lock.json artifacts/package-lock.json
+
+      - name: Clone Flathub Shared Modules
+        working-directory: build
+        run: git clone --depth 1 https://github.com/flathub/shared-modules.git
```
Comment (ReenigneArcher) [inline .github/workflows/CI.yml:228]: Use github checkout action

File: .github/workflows/CI.yml [hunk 5]
```diff
@@ -218,20 +213,46 @@ jobs:
...
 
+      - name: Generate cache for npm
+        run: |
+          npm install --lockfile-version 2 --package-lock-only
+          git clone --depth 1 https://github.com/flatpak/flatpak-builder-tools.git
+          pip install ./flatpak-builder-tools/node
+          flatpak-node-generator -o npm-generated-sources.json npm package-lock.json
+          cp npm-generated-sources.json build/npm-generated-sources.json
+          mv npm-generated-sources.json artifacts/npm-generated-sources.json
+          mv package-lock.json artifacts/package-lock.json
```
Comment (ReenigneArcher) [inline .github/workflows/CI.yml:224]: Everything in `./artifacts` is a release asset... please use another directory.

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]

File: .github/workflows/CI.yml [hunk 6]
```diff
@@ -218,20 +213,46 @@ jobs:
...
-          sudo su $(whoami) -c 'flatpak build-bundle --arch=${{ matrix.arch }} ./repo \
-            ../artifacts/sunshine_${{ matrix.arch }}.flatpak dev.lizardbyte.sunshine'
-          sudo su $(whoami) -c 'flatpak build-bundle --runtime --arch=${{ matrix.arch }} ./repo \
-            ../artifacts/sunshine_debug_${{ matrix.arch }}.flatpak dev.lizardbyte.sunshine.Debug'
+
+      - name: Bundle Flatpak
+        run: |
+          flatpak repo --branches build/local-repo
+          flatpak build-bundle build/local-repo --arch=${{ matrix.arch }} artifacts/sunshine_${{ matrix.arch }}.flatpak \
+            dev.lizardbyte.sunshine master
```
[the remark this answers is not in the document]
Comment (istori1) [reply .github/workflows/CI.yml:261]: Flatpak uses branches like git. The default branch is _master_. When Flathub build an app they use _stable_ as the default branch, for PRs they use _testing_ as the branch. This is to ensure that we use _master_ to differentiate from the Flathub build.

https://docs.flatpak.org/en/latest/under-the-hood.html?highlight=branch#git-for-apps

https://docs.flatpak.org/en/latest/single-file-bundles.html?highlight=branch#single-file-bundles

[the remark this answers is not in the document]
Comment (ReenigneArcher) [reply .github/workflows/CI.yml:261]: Shouldn't we use the actual branch name or PR number?

File: src_assets/linux/assets/apps.json
Comment (ReenigneArcher) [inline src_assets/linux/assets/apps.json:1]: Let's move this change to a separate PR.

File: packaging/linux/flatpak/dev.lizardbyte.sunshine.yml [hunk 7]
```diff
@@ -1,5 +1,5 @@
 ---
-app-id: dev.lizardbyte.sunshine
+app-id: dev.lizardbyte.sunshine # Flathub build is dev.lizardbyte.app.Sunshine
```
[the remark this answers is not in the document]
Comment (istori1) [reply packaging/linux/flatpak/dev.lizardbyte.sunshine.yml]: ID's should be different to prevent conflict between builds.

File: .github/workflows/CI.yml [hunk 8]
```diff
@@ -218,20 +213,68 @@ jobs:
...
-          sudo su $(whoami) -c 'flatpak build-bundle --arch=${{ matrix.arch }} ./repo \
-            ../artifacts/sunshine_${{ matrix.arch }}.flatpak dev.lizardbyte.sunshine'
-          sudo su $(whoami) -c 'flatpak build-bundle --runtime --arch=${{ matrix.arch }} ./repo \
-            ../artifacts/sunshine_debug_${{ matrix.arch }}.flatpak dev.lizardbyte.sunshine.Debug'
+
+      - name: Bundle Flatpak
+        run: |
+          flatpak repo --branches build/local-repo
+          flatpak build-bundle build/local-repo --arch=${{ matrix.arch }} artifacts/sunshine_${{ matrix.arch }}.flatpak \
+            dev.lizardbyte.sunshine ${{ github.ref_name }}
```
Comment (ReenigneArcher) [inline .github/workflows/CI.yml:269]: I'd prefer not to use the `${{ github.ref_name }}` as it's not consistent between PRs and push events.

File: .github/workflows/CI.yml [hunk 9]
```diff
@@ -218,20 +213,68 @@ jobs:
...
         run: |
-          sudo su $(whoami) -c 'flatpak run org.flatpak.Builder --arch=${{ matrix.arch }} --repo=repo --force-clean \
-            --stop-at=cuda build-sunshine dev.lizardbyte.sunshine.yml'
+          mkdir -p ~/.local/bin
+          wget https://github.com/mikefarah/yq/releases/latest/download/yq_linux_amd64 -O ~/.local/bin/yq
+          chmod +x ~/.local/bin/yq
+          # Replace value for filesystem permission
+          yq -i '.modules[-2].build-options.build-args[0] = "--filesystem=${{ github.workspace }}"' dev.lizardbyte.sunshine.yml
+          # Replace value for WORKSPACE variable
+          yq -i '.modules[-2].build-options.env.WORKSPACE = "${{ github.workspace }}"' dev.lizardbyte.sunshine.yml
```
Comment (ReenigneArcher) [inline .github/workflows/CI.yml:251]: This yq tool seems pretty cool, but I have a concern/question with replacing values like this.

Currently we're replacing values using cmake and the configure command. There are already 3 values replaced just for flatpak. I'm thinking it should be done in a similar way for these as to not complicate an offline build.

Can just modify this step above...
```bash
          cmake -DGITHUB_CLONE_URL=${clone_url} \
            -DGITHUB_BRANCH=${branch} \
            -DGITHUB_COMMIT=${commit} \
            -DSUNSHINE_CONFIGURE_FLATPAK_MAN=ON \
            -DSUNSHINE_CONFIGURE_ONLY=ON \
            ..
```

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]

**Comment by ReenigneArcher** [comment]
I want to get to parity with the flathub version, but can this be broken out into multiple (incremental PRs)?

It's a bit overwhelming as is. I think first PR should just change the ubuntu packages with upstream packages, then we can go from there.
